pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/functioncache/FunctionCacheManagerImpl.java

```diff
@@ -73,39 +71,37 @@ public void registerFunctionInstance(String fid,
             throw new NullPointerException("FunctionID not set");
         }
 
-        synchronized (cacheFunctions) {
-            FunctionCacheEntry entry = cacheFunctions.get(fid);
+        FunctionCacheEntry entry = cacheFunctions.get(fid);
 
-            if (null == entry) {
-                URL[] urls = new URL[requiredJarFiles.size() + requiredClasspaths.size()];
-                int count = 0;
-                try {
-                    // add jar files to urls
-                    for (String jarFile : requiredJarFiles) {
-                        urls[count++] = new File(jarFile).toURI().toURL();
-                    }
+        if (null == entry) {
+            URL[] urls = new URL[requiredJarFiles.size() + requiredClasspaths.size()];
+            int count = 0;
+            try {
+                // add jar files to urls
+                for (String jarFile : requiredJarFiles) {
+                    urls[count++] = new File(jarFile).toURI().toURL();
+                }
 
-                    // add classpaths
-                    for (URL url : requiredClasspaths) {
-                        urls[count++] = url;
-                    }
+                // add classpaths
+                for (URL url : requiredClasspaths) {
+                    urls[count++] = url;
+                }
 
-                    cacheFunctions.put(
+                cacheFunctions.put(
                         fid,
                         new FunctionCacheEntry(
-                            requiredJarFiles,
-                            requiredClasspaths,
-                            urls,
-                            eid, rootClassLoader));
-                } catch (Throwable cause) {
-                    Exceptions.rethrowIOException(cause);
-                }
```

Comment:
We should use `computeIfAbsent` here. Because in the `if` branch, the composed `get` and `put` operations are not atomic.

We can use a `AtomicBoolean` to check if a new value has been put, like:

```java
        final Map<String, String> map = new ConcurrentHashMap<>();
        final AtomicBoolean flag = new AtomicBoolean(false);
        final String value = map.computeIfAbsent("A", __ -> {
            flag.set(true);
            return "B";
        });
```